zallet/src/components/json_rpc/methods/recover_accounts.rs

```diff
@@ -0,0 +1,144 @@
+use std::collections::HashMap;
+
+use jsonrpsee::{
+    core::RpcResult,
+    types::{ErrorCode as RpcErrorCode, ErrorObjectOwned},
+};
+use serde::{Deserialize, Serialize};
+use zaino_state::fetch::FetchServiceSubscriber;
+use zcash_client_backend::{
+    data_api::{Account as _, AccountBirthday, WalletRead, WalletWrite},
+    proto::service::TreeState,
+};
+use zcash_protocol::consensus::{BlockHeight, NetworkType, Parameters};
+use zip32::fingerprint::SeedFingerprint;
+
+use crate::components::{
+    database::DbConnection,
+    json_rpc::{server::LegacyCode, utils::ensure_wallet_is_unlocked},
+    keystore::KeyStore,
+};
+
+/// Response to a `z_recoveraccounts` RPC request.
+pub(crate) type Response = RpcResult<Accounts>;
+
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct AccountParameter<'a> {
+    name: &'a str,
+    seedfp: &'a str,
+    zip32_account_index: u32,
+    birthday_height: u32,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub(crate) struct Accounts {
+    accounts: Vec<Account>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct Account {
+    /// The account's UUID within this Zallet instance.
+    account_uuid: String,
+
+    seedfp: String,
+
+    /// The account's ZIP 32 account index.
+    zip32_account_index: u32,
+}
+
+pub(crate) async fn call(
+    wallet: &mut DbConnection,
+    keystore: &KeyStore,
+    chain: FetchServiceSubscriber,
+    accounts: Vec<AccountParameter<'_>>,
+) -> Response {
+    ensure_wallet_is_unlocked(keystore).await?;
+    // TODO: Ensure wallet is backed up.
+
+    let recover_until = wallet
+        .chain_height()
+        .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?
+        .ok_or(LegacyCode::InWarmup.with_static("Wallet sync required"))?;
+
+    // Prepare arguments for the wallet.
+    let mut account_args = vec![];
+    for account in accounts {
+        let seed_fp = {
+            let mut hash = [0; 32];
+            hex::decode_to_slice(account.seedfp, &mut hash).map_err(|e: hex::FromHexError| {
+                LegacyCode::InvalidParameter.with_message(format!("Invalid seed fingerprint: {e}"))
+            })?;
+            SeedFingerprint::from_bytes(hash)
+        };
+
+        let account_index =
+            zip32::AccountId::try_from(account.zip32_account_index).map_err(|e| {
+                LegacyCode::InvalidParameter
+                    .with_message(format!("Invalid ZIP 32 account index: {e}"))
+            })?;
+
+        let birthday_height = BlockHeight::from_u32(account.birthday_height);
+
+        let treestate = {
```

Comment:
This code now exists in a few different places in the node, we should factor it out.